aplbrain/neuvue-app
# Self assign tasks

Add ability for user to self assign tasks under specific conditions.
* User must be in "Can self assign tasks" group
* User's number of currently assigned tasks must be fewer than the maximum number of tasks allowed for a namespace

If user tries to add tasks and there are none left in the unassigned bucket, a toast message will let them know. Please let me know if you find any bugs or have any thoughts on styling and user experience.

Changes:
* Add "Add more tasks" button to namespace headers on tasks page. Button will only be visible for a namespace if the top two conditions are satisfied.
* Add `number_of_tasks_users_can_self_assign` and `max_number_of_pending_tasks_per_user` to namespace model
* Add "Can self assign tasks" group
* Add loading spinner to button
* Add post request for `/tasks` and attach it to "Add more tasks" button

Tests suggested:
* Log into the django admin dashboard in another window so that you can test settings
* Navigate to the /tasks page 
  * Confirm that no "Add More Tasks" buttons are visible
* Add yourself to the "Can self assign tasks" group
  * Confirm that "Add More Tasks" buttons are visible
* Lower the max tasks limit for one of the namespaces to the number of tasks you are currently assigned
  * Confirm that the "Add More Tasks" button disappears for that namespace
* Raise the limit back up. Start a second Neuvue client instance and tally the unassigned tasks for your namespaces. Keep the client instance open.
```
from neuvueclient import NeuvueQueue
q = NeuvueQueue("https://queue.neuvue.io")
tasks = q.get_tasks(sieve={"assignee": "unassigned", "namespace": <namespace>, "status":["open", "pending"]}, return_states=False, return_metadata=False)
```
* Click the "Add More Tasks" button for a namespace that has at least one unassigned task.
  * Confirm that the correct number of tasks is added. Unassign them from yourself.
```
ids = tasks.index.tolist()
for id in ids:
  q.patch_task(id, assignee="unassigned")
```
* Click the "Add More Tasks" button for a namespace that has no unassigned tasks.
  * Confirm that a toast message pops up with a warning message, and that the loading spinner is removed from the button.
* Change the max tasks limit for a namespace to be 1 greater than the number of tasks you are currently assigned. Click the "Add More Tasks" button.
  * Confirm that only 1 new task is assigned to you and that the "Add More Tasks" button disappears. Unassign the task from yourself.
* Any other tests you can think of!
